machine-learning/ml/pipelines/binary.py

```diff
@@ -39,48 +65,82 @@ def run(self):
-                    continue
-
-                # we split in train and test
-                # (note that we use the same split for all the models)
-                x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=TEST_SPLIT_SIZE, random_state=42)
-
-                for model in self._models_to_run:
-                    model_name = model.name()
-
-                    try:
-                        log("Model {}".format(model.name()))
-                        self._start_time()
-                        test_scores, model_to_save = self._run_single_model(model, x, y, x_train, x_test, y_train, y_test)
-
-                        # log test scores
-                        log(format_test_results(dataset, refactoring_name, model_name + " test", test_scores["precision"],
-                                           test_scores["recall"], test_scores['accuracy'], test_scores['tn'],
-                                           test_scores['fp'], test_scores['fn'], test_scores['tp']))
-
-                        # we save the best estimator we had during the search
-                        model.persist(dataset, refactoring_name, features, model_to_save, scaler)
-
-                        self._finish_time(dataset, model, refactoring)
-                    except Exception as e:
-                        print(e)
-                        print(str(traceback.format_exc()))
-
-                        log("An error occurred while working on refactoring " + refactoring_name + " model " + model.name())
-                        log(str(e))
-                        log(str(traceback.format_exc()))
-
-
-    def _run_single_model(self, model_def, x, y, x_train, x_test, y_train, y_test):
+                # we have two options to select a test set,
+                # 1.) Predefined in the database
+                if TEST_SPLIT_SIZE < 0 and len(VALIDATION_DATASETS) > 0:
+                    train_features, x_train, y_train, scaler = retrieve_labelled_instances(dataset, refactoring, True)
+                    # test if any refactorings were found for the given refactoring type
+                    if x_train is None:
+                        log("Skip model building for refactoring type: " + refactoring.name())
+                        continue
+
+                    x_tests, y_tests, dataset_name = [], [], []
+                    for validation_dataset in VALIDATION_DATASETS:
+                        dataset_name.append(validation_dataset)
+                        test_features, x_test, y_test, _ = retrieve_labelled_instances(validation_dataset, refactoring, False, train_features)
+                        # test if any refactorings were found for the given refactoring type
+                        if x_test is None:
+                            log("Skip model building for refactoring type: " + refactoring.name())
+                            continue
+                        if len(x_tests) == 0:
```

Comment:
Do you need this if? Can't you use the `append` directly?

Reply:
I will try it

Reply:
Works

Reply:
But I still see the if here

Can't it just be

```
x_tests.append(x_test)
y_tests.append(y_test)
```

without the if?

Reply:
It got last in the merge, I fixed that.